[export] Upstream unflattener.

zhxchen17 opened the pull request:
Summary: Provide a way for users to get the original module structure back after exporting.

Test Plan: caffe2/test:test_export -- -r unflatten

Differential Revision: D50708490




cc @avikchaudhuri @gmagogsfm @tugsbayasgalan @angelayi

suo commented on test/export/test_export.py:
```diff
@@ -109,7 +109,163 @@ def f(x, y):
+        unflattened_output = unflattened(*args)
+        self.assertTrue(torch.allclose(orig_output, unflattened_output))
+
+    def test_unflatten_nested(self):
```
can we make this a separate test file? The unflattener is a logically separate component from export, and it would be nice to avoid bloating this file over time.

zhxchen17 committed:
[export] Upstream unflattener. (#112189)

Summary:

Provide a way for users to get the original module structure back after exporting.

Test Plan: caffe2/test:test_export -- -r unflatten

Reviewed By: suo, angelayi

Differential Revision: D50708490

zhxchen17 commented on torch/_export/unflatten.py:
```diff
@@ -0,0 +1,636 @@
+    We want to inplace this into something that looks like the original eager code:
+        def forward(self, buffer, x):
+            mutated_buffer = aten.add(buffer, x)
+            buffer.copy_(mutated_buffer)
```
If we have a repro we can open an issue for this.

zhxchen17 commented:
@pytorchbot merge
